File: myq2.py
import sys

def init():
  global value
  value={}
  value[' ']=32
  a,A=ord('a'),ord('A')
  for i in range(26):
    value[chr(a+i)]=i+1
    value[chr(A+i)]=i+51

def populate(n,m):
  global s, value, val, ind
  val[0]=value[s[0]]%m;
  ind[val[0]]=0

  for i in range(1,n):
    val[i]=(val[i-1]+value[s[i]])%m
    ind[val[i]]=i

def main():
  global val,ind,s

  inp=sys.stdin
  line=inp.readline()
  t=int(line.strip())
  init()
  for i in range(t):
    s=inp.readline()[:-1]
    line=inp.readline()
    m,k=line.strip().split()
    m,k=int(m),int(k)

    k=k%m
    n=len(s)
    val=[0]*n
    ind=[-1]*m
    populate(n,m)

    if ind[k]==-1:
      mind,mlen=-1,-1
    else:
      mind,mlen=0,ind[k]+1

    tmp,tlen=0,0

    for i in range(1,n):
      tmp=(k+val[i-1])%m
      if ind[tmp]==-1: continue

      tlen=ind[tmp]-i+1
      if tlen>mlen:
        mind,mlen=i,tlen
    
    if mind!=-1:
      print mind,mlen
    else:
      print mind

if __name__=='__main__':
   main()

File: myq2.cpp
#include <stdio.h>
#include <string.h>

using namespace std;

#define N 1000001
#define M 1000008
int m, k;
int val[N];
int ind[M];
char* s;

void init(int n)
{
  int i=0;
  for(i=0;i<m;++i) ind[i]=-1;

  if(' '==s[0]) val[0]=32%m;
  else if('a'<=s[0] && s[0]<='z') val[0]=(s[0]-'a'+1)%m;
  else val[0]=(s[0]-'A'+51)%m;
  ind[val[0]]=0;

  for(i=1;i<n;++i)
  {
    if(' '==s[i]) val[i]=(val[i-1]+32)%m;
    else if('a'<=s[i] && s[i]<='z') val[i]=(val[i-1]+(s[i]-'a'+1))%m;
    else val[i]=(val[i-1]+(s[i]-'A'+51))%m;
    ind[val[i]]=i;
  }
}

int main()
{
  int t=0;
  int i=0,n=0;
  scanf("%d",&t);
  s=new char[N];
  char c; int ssize=0;
  while(t--)
  {
    getchar();//read \n from previous
    fgets(s,N,stdin);

    n=strlen(s);
    s[n-1]='\0';
    n-=1;
    scanf("%d %d",&m,&k);

    init(n);    
    k=(m+k%m)%m;
    int len=ind[k]+1; //starting at 0
    int tmp, tlen=len;
    int mind=0, mlen=len;
    if(!len) { mind=-1; mlen=-1;}

    for(i=1;i<n;++i)
    {
      tmp=(k+val[i-1])%m;
      if(ind[tmp]==-1) continue;
      tlen=ind[tmp]-i+1;
      if(tlen>mlen)
      {
        mlen=tlen;
	mind=i;
      }
    }
/*
    for(i=0;i<n;++i) printf("%d ",val[i]);
    printf("\n");
    for(i=0;i<m;++i) printf("%d ",ind[i]);
    printf("\n");
    printf("%s %d\n",s,n);
*/
    if(mind==-1)
      printf("-1\n");
    else
      printf("%d %d\n",mind,mlen);
  }
  return 0;
} 
